File: calcity_app/lib/services/weather_service.dart
import 'dart:convert';

import 'package:flutter/material.dart';
import 'package:http/http.dart' as http;

/// Live weather for California City, CA — fetched directly from
/// Open-Meteo (free, no API key) so conditions update through the day.
class HourlyForecast {
  final DateTime time;
  final int temp;
  final int code;
  final int precipProb;

  const HourlyForecast({
    required this.time,
    required this.temp,
    required this.code,
    required this.precipProb,
  });
}

class LiveWeather {
  final int temp; // °F now
  final int feelsLike; // °F apparent
  final int humidity; // %
  final double windMph;
  final bool isDay;
  final int conditionCode; // WMO code
  final int highToday;
  final int lowToday;
  final String sunrise;
  final String sunset;
  final List<HourlyForecast> hourly;
  final DateTime fetchedAt;

  const LiveWeather({
    required this.temp,
    required this.feelsLike,
    required this.humidity,
    required this.windMph,
    required this.isDay,
    required this.conditionCode,
    required this.highToday,
    required this.lowToday,
    required this.sunrise,
    required this.sunset,
    required this.hourly,
    required this.fetchedAt,
  });

  String get condition => _wmoInfo(conditionCode, isDay).$2;

  IconData get icon => wmoIcon(conditionCode, isDay);

  /// Public WMO-code → icon mapper (used by widgets).
  static IconData wmoIcon(int code, bool isDay) => _wmoInfo(code, isDay).$1;

  /// Hourly strip starting at the current hour (up to [count] entries).
  List<HourlyForecast> nextHours(int count) {
    final now = DateTime.now();
    final start = hourly.indexWhere((h) =>
        h.time.isAfter(now.subtract(const Duration(minutes: 30))));
    final from = start < 0 ? 0 : start;
    return hourly.skip(from).take(count).toList();
  }

  factory LiveWeather.fromJson(Map<String, dynamic> json) {
    final current = (json['current'] ?? {}) as Map<String, dynamic>;
    final hourlyRaw = (json['hourly'] ?? {}) as Map<String, dynamic>;
    final dailyRaw = (json['daily'] ?? {}) as Map<String, dynamic>;

    final times = (hourlyRaw['time'] as List?) ?? [];
    final temps = (hourlyRaw['temperature_2m'] as List?) ?? [];
    final codes = (hourlyRaw['weather_code'] as List?) ?? [];
    final precipp = (hourlyRaw['precipitation_probability'] as List?) ?? [];

    final hourly = <HourlyForecast>[];
    for (var i = 0; i < times.length; i++) {
      hourly.add(HourlyForecast(
        time: DateTime.parse(times[i] as String),
        temp: (temps[i] as num?)?.round() ?? 0,
        code: (codes[i] as num?)?.toInt() ?? 0,
        precipProb: (precipp[i] as num?)?.round() ?? 0,
      ));
    }

    final dailyTimes = (dailyRaw['time'] as List?) ?? [];
    final dailyMax = (dailyRaw['temperature_2m_max'] as List?) ?? [];
    final dailyMin = (dailyRaw['temperature_2m_min'] as List?) ?? [];
    final sunrises = (dailyRaw['sunrise'] as List?) ?? [];
    final sunsets = (dailyRaw['sunset'] as List?) ?? [];

    int? high, low;
    String sunrise = '', sunset = '';
    final nowDate = DateTime.now();
    for (var i = 0; i < dailyTimes.length; i++) {
      final d = DateTime.parse(dailyTimes[i] as String);
      if (d.year == nowDate.year && d.month == nowDate.month && d.day == nowDate.day) {
        high = (dailyMax[i] as num?)?.round();
        low = (dailyMin[i] as num?)?.round();
        final sr = (sunrises[i] as String?);
        final ss = (sunsets[i] as String?);
        if (sr != null && sr.isNotEmpty) {
          sunrise = _formatTime(DateTime.parse(sr));
        }
        if (ss != null && ss.isNotEmpty) {
          sunset = _formatTime(DateTime.parse(ss));
        }
      }
    }

    return LiveWeather(
      temp: (current['temperature_2m'] as num?)?.round() ?? 0,
      feelsLike: (current['apparent_temperature'] as num?)?.round() ?? 0,
      humidity: (current['relative_humidity_2m'] as num?)?.round() ?? 0,
      windMph: (current['wind_speed_10m'] as num?)?.toDouble() ?? 0,
      isDay: (current['is_day'] as num?)?.toInt() == 1,
      conditionCode: (current['weather_code'] as num?)?.toInt() ?? 0,
      highToday: high ?? 0,
      lowToday: low ?? 0,
      sunrise: sunrise,
      sunset: sunset,
      hourly: hourly,
      fetchedAt: DateTime.now(),
    );
  }

  static String _formatTime(DateTime dt) {
    var h = dt.hour;
    final min = dt.minute.toString().padLeft(2, '0');
    final ampm = h >= 12 ? 'PM' : 'AM';
    h = h % 12;
    if (h == 0) h = 12;
    return '$h:$min $ampm';
  }

  /// WMO weather code → (icon, short label).
  static (IconData, String) _wmoInfo(int code, bool isDay) {
    if (code == 0) {
      return isDay
          ? (Icons.wb_sunny_rounded, 'Clear')
          : (Icons.nightlight_round, 'Clear');
    }
    if (code == 1) {
      return isDay
          ? (Icons.wb_sunny_outlined, 'Mostly clear')
          : (Icons.nights_stay_outlined, 'Mostly clear');
    }
    if (code == 2) return (Icons.cloud_queue_rounded, 'Partly cloudy');
    if (code == 3) return (Icons.cloud_rounded, 'Overcast');
    if (code == 45 || code == 48) return (Icons.foggy, 'Fog');
    if (code >= 51 && code <= 57) return (Icons.grain_rounded, 'Drizzle');
    if (code >= 61 && code <= 67) return (Icons.water_drop_rounded, 'Rain');
    if (code >= 71 && code <= 77) return (Icons.ac_unit_rounded, 'Snow');
    if (code >= 80 && code <= 82) return (Icons.umbrella_rounded, 'Showers');
    if (code >= 85 && code <= 86) return (Icons.ac_unit_rounded, 'Snow showers');
    if (code == 95) return (Icons.thunderstorm_rounded, 'Thunderstorm');
    if (code >= 96) return (Icons.thunderstorm_rounded, 'Storm');
    return (Icons.cloud_rounded, 'Unknown');
  }
}

class WeatherService {
  WeatherService._();

  static const String _base = 'https://api.open-meteo.com/v1/forecast';

  /// California City, CA (35.1258, -117.9859).
  static const double _lat = 35.1258;
  static const double _lon = -117.9859;

  /// Fetches current + hourly + daily weather. Returns null on any failure
  /// (offline, timeout, malformed) — callers fall back to their last value.
  static Future<LiveWeather?> fetch() async {
    final url = Uri.parse(
        '$_base?latitude=$_lat&longitude=$_lon'
        '&current=temperature_2m,relative_humidity_2m,apparent_temperature,is_day,weather_code,wind_speed_10m'
        '&hourly=temperature_2m,weather_code,precipitation_probability'
        '&daily=weather_code,temperature_2m_max,temperature_2m_min,sunrise,sunset'
        '&temperature_unit=fahrenheit&wind_speed_unit=mph&precipitation_unit=inch'
        '&timezone=America%2FLos_Angeles&forecast_days=2');
    try {
      final resp = await http.get(url).timeout(const Duration(seconds: 12));
      if (resp.statusCode != 200) return null;
      return LiveWeather.fromJson(
          json.decode(resp.body) as Map<String, dynamic>);
    } catch (_) {
      return null; // offline or timeout — keep showing the last known
    }
  }
}

File: community/services/weather_service.py
"""Live weather for the management dashboard — Open-Meteo, no API key.

The mobile app fetches Open-Meteo directly; this service gives the Django
dashboard the same live data (current conditions + today's forecast) so the
dashboard weather changes through the day instead of only when someone
manually posts an update.

Fetches are cached for 20 minutes (per-process LocMemCache is fine here —
the dashboard is low-traffic; worst case is one extra fetch per process).
Manual advisory posts (fire risk, heat advisories) live on WeatherInfo and
are shown alongside.
"""
import json
from datetime import datetime
from urllib.request import urlopen

from django.core.cache import cache

LAT, LON = 35.1258, -117.9859  # California City, CA
CACHE_KEY = "calcity_live_weather"
CACHE_TTL = 20 * 60  # 20 minutes

# WMO weather codes -> (label, emoji)
WMO = {
    0: ("Clear", "\u2600\ufe0f"),
    1: ("Mostly clear", "\u2600\ufe0f"),
    2: ("Partly cloudy", "\u26c5"),
    3: ("Overcast", "\u2601\ufe0f"),
    45: ("Fog", "\ud83c\udf2b\ufe0f"),
    48: ("Fog", "\ud83c\udf2b\ufe0f"),
    51: ("Drizzle", "\ud83c\udf27\ufe0f"),
    53: ("Drizzle", "\ud83c\udf27\ufe0f"),
    55: ("Drizzle", "\ud83c\udf27\ufe0f"),
    61: ("Rain", "\ud83c\udf27\ufe0f"),
    63: ("Rain", "\ud83c\udf27\ufe0f"),
    65: ("Rain", "\ud83c\udf27\ufe0f"),
    80: ("Showers", "\ud83c\udf27\ufe0f"),
    81: ("Showers", "\ud83c\udf27\ufe0f"),
    82: ("Showers", "\ud83c\udf27\ufe0f"),
    95: ("Thunderstorm", "\u26c8\ufe0f"),
    96: ("Storm", "\u26c8\ufe0f"),
    99: ("Storm", "\u26c8\ufe0f"),
}

API_URL = (
    "https://api.open-meteo.com/v1/forecast"
    f"?latitude={LAT}&longitude={LON}"
    "&current=temperature_2m,relative_humidity_2m,apparent_temperature,is_day,weather_code,wind_speed_10m"
    "&daily=temperature_2m_max,temperature_2m_min,sunrise,sunset"
    "&temperature_unit=fahrenheit&wind_speed_unit=mph"
    "&timezone=America%2FLos_Angeles&forecast_days=1"
)


def _fmt_time(iso):
    try:
        dt = datetime.fromisoformat(iso)
    except (TypeError, ValueError):
        return ""
    h = dt.hour % 12 or 12
    return f"{h}:{dt.minute:02d} {'PM' if dt.hour >= 12 else 'AM'}"


def _fetch():
    with urlopen(API_URL, timeout=12) as resp:
        data = json.loads(resp.read())
    cur = data.get("current", {})
    daily = data.get("daily", {})
    code = int(cur.get("weather_code", 0))
    label, emoji = WMO.get(code, ("Unknown", "\u2753"))
    return {
        "temp": round(cur.get("temperature_2m", 0)),
        "feels_like": round(cur.get("apparent_temperature", 0)),
        "condition": label,
        "emoji": emoji,
        "humidity": cur.get("relative_humidity_2m"),
        "wind_mph": round(cur.get("wind_speed_10m", 0)),
        "is_day": cur.get("is_day") == 1,
        "high": round(daily.get("temperature_2m_max", [0])[0]),
        "low": round(daily.get("temperature_2m_min", [0])[0]),
        "sunrise": _fmt_time(daily.get("sunrise", [""])[0]),
        "sunset": _fmt_time(daily.get("sunset", [""])[0]),
        "updated": datetime.now(),
    }


def get_live_weather(force=False):
    """Return a dict of live weather, or None if the fetch fails.

    Cached for CACHE_TTL; pass force=True to bypass the cache.
    """
    if not force:
        cached = cache.get(CACHE_KEY)
        if cached:
            return cached
    try:
        data = _fetch()
    except Exception:
        return None
    cache.set(CACHE_KEY, data, CACHE_TTL)
    return data
